support for gpu on windows

guschmue opened the pull request:
code changes to enable gpu support for windows

Commit:
gpu support on windows

Commit:
code changes to enable gpu support for windows
Merge remote-tracking branch 'schmuell/windows-gpu-0' into windows-gpu-0

tensorflow-jenkins commented:
Can one of the admins verify this patch?

mrry commented on tensorflow/core/framework/allocator.h:
```diff
@@ -152,6 +152,7 @@ class Allocator {
   // allocated by this allocator.
   virtual size_t RequestedSize(void* ptr) {
     CHECK(false) << "allocator doesn't track sizes";
+    return size_t(0);
```
Is this necessary? `LogMessageFatal::~LogMessageFatal()` has `TF_ATTRIBUTE_NORETURN`, so the lack of a return statement shouldn't (AIUI) give an error. Or is this compiler-version-specific?

(Same goes for the other similar changes in here.)

guschmue replied:
it is only needed because msvc fails debug builds that the function has no return value

mrry replied:
It works in Release/RelWithDebInfo builds though, right?

I'd prefer to keep these as-is for now, as I think there are other issues that prevent a Debug build from succeeding (something about taking the address of zero-length iterators, if my memory serves me). Can you split these changes out and we'll consider them separately?

mrry commented on tensorflow/stream_executor/cuda/cuda_driver.cc:
```diff
@@ -38,6 +38,14 @@ limitations under the License.
 #include "tensorflow/stream_executor/lib/inlined_vector.h"
 
+#if defined(PLATFORM_WINDOWS)
+// FIXME: in windows ARRAYSIZE is defined in winnt.h but including it
```
Nit: The string `FIXME` gets intercepted by one of the linters we use internally. Can you please change it to `TODO`? Thanks!

guschmue replied:
ok

mrry commented on tensorflow/stream_executor/cuda/cuda_rng.cc:
```diff
@@ -15,8 +15,9 @@ limitations under the License.
+#if !defined(PLATFORM_WINDOWS)
 #include <dlfcn.h>
-
+#endif
```
I think this is a stray `#include` that is no longer used, and we can just delete it.

guschmue replied:
ok

mrry reviewed (changes requested):
Looking good thanks. Just minor changes needed - I think we don't need to add the return statements as they're (now) covered by a MSVC-recognized attribute in `LogMessageFatal::~LogMessageFatal()`. Can you check whether it works without those changes?

vrv commented on tensorflow/core/common_runtime/gpu/gpu_device.cc:
```diff
@@ -882,7 +882,13 @@ Status BaseGPUDeviceFactory::GetValidDeviceIds(
   if (visible_device_list.empty()) {
     visible_gpu_order.resize(gpu_manager->VisibleDeviceCount());
     // By default, visible to virtual mapping is unchanged.
+#if defined(PLATFORM_WINDOWS)
```
if std::iota isn't available on windows but std::generate works on all platforms, perhaps we just use the std::generate version and avoid the ifdef

guschmue replied:
makes sense. I can make sure its fine on linux.

Commit:
changes after review:
cuda_driver.cc: changed FIXME to TODO
cuda_rng.cc: removed stray include
gpu_device.cc: use  std::generate() for all platforms to avoid ifdef

when building for linux I found:
load_library.h: added FormatLibraryFileName() back - linux still needs it

guschmue commented:
pushed the requested changes

mrry reviewed (commented):
LGTM. It'd be good if you could still remove the extra return statements after `LOG(FATAL)`/`CHECK(false)`, because these are likely to bit-rot (we don't test the Debug build on Windows).

mrry commented:
@tensorflow-jenkins test this please.

JimSEOW commented:
**For people looking for ways to compile:**

It is in
https://github.com/schmuell/tensorflow/blob/b11ef5128f7341cfc7018ac0771d8939b8535cfc/tensorflow/contrib/cmake/README.md

you need to instal the cuda 8.0 SDK + cudnn 5.1 and add 2 options to cmake
-Dtensorflow_ENABLE_GPU=ON  -DCUDNN_HOME=c:\local\cudnn

I normally do:
cd tensorflow\contrib\cmake\
mkdir build
cd build
cmake .. -A x64 -DCMAKE_BUILD_TYPE=RelWithDebInfo
-DPYTHON_EXECUTABLE=C:\local\Anaconda2\envs\py3\python.exe
-DPYTHON_LIBRARIES=C:\local\Anaconda2\envs\py3\libs\python35.lib
-DSWIG_EXECUTABLE=c:\local\swigwin-3.0.10\swig.exe
-Dtensorflow_ENABLE_GPU=ON  -DCUDNN_HOME="C:\local\cudnn"
msbuild /p:Configuration=RelWithDebInfo tf_python_build_pip_package.vcxproj

Carmezim commented:
@mrry Thank you for your awesome contributions! Should the [readme]https://github.com/schmuell/tensorflow/blob/b11ef5128f7341cfc7018ac0771d8939b8535cfc/tensorflow/contrib/cmake/README.md) be updated?

Carmezim commented:
@mrry @guschmue awesome contribution, should the readme be updated as it is confusing some folks due to:
'Currently, only CPU builds are supported, but we are working on
 providing a GPU build as well.'?

guschmue commented:
Fair, we can fix that and add some more details to the readme file. 
Should be getting to it today.

cerrayo commented:
Following the above build instructions, I am getting the following errors when trying to compile:

"C:\ST\Tensorflow\tensorflow\tensorflow\contrib\cmake\build\tf_python_build_pip_package.vcxproj" (default target) (1) ->
"C:\ST\Tensorflow\tensorflow\tensorflow\contrib\cmake\build\pywrap_tensorflow.vcxproj" (default target) (3) ->
"C:\ST\Tensorflow\tensorflow\tensorflow\contrib\cmake\build\tf_core_gpu_kernels.vcxproj" (default target) (33) ->
(CustomBuild target) ->
  c:\st\tensorflow\tensorflow\tensorflow\contrib\cmake\build\external\eigen_archive\eigen\src/Core/Ref.h(89): error : "operator=" has already been declared in the current scope [C:\ST\Tensorflow\tensorflow\tensorflow\contrib\cmake\build
\tf_core_gpu_kernels.vcxproj]
  c:\st\tensorflow\tensorflow\tensorflow\contrib\cmake\build\external\eigen_archive\eigen\src/Core/Block.h(111): error : "operator=" has already been declared in the current scope [C:\ST\Tensorflow\tensorflow\tensorflow\contrib\cmake\bu
ild\tf_core_gpu_kernels.vcxproj]

    369 Warning(s)
    2 Error(s)

This is with the following cmake command:

cmake .. -A x64 -DCMAKE_BUILD_TYPE=Release 
-DSWIG_EXECUTABLE=C:/ST/Tensorflow/swigwin-3.0.10/swig.exe 
-Dtensorflow_ENABLE_GPU=ON 
-DCUDNN_HOME=C:/ST/Tensorflow/cuda8_0_cudnn5_1 
-DPYTHON_EXECUTABLE=C:/Anaconda3/envs/tensorflow/python.exe 
-DPYTHON_LIBRARIES=C:/Anaconda3/envs/tensorflow/libs/python35.lib

And the following MSBuild command:

MSBuild /p:Configuration=Release tf_python_build_pip_package.vcxproj

guschmue commented:
yes, see here:
https://github.com/tensorflow/tensorflow/pull/5076
Rhis should resolve once the eigen version in the master get updated.
If you want it fixed right away you could apply 5076.

Carmezim commented:
@guschmue All working great so far, thank you for this contribution. 
If you want I can make those edits on "current limitations" and "current status" regarding the availability for GPU support and submit a PR, just ping me if you will.

karthiek commented:
@JimSEOW Hi

I tried to build following the steps above. 
I was able to build with Dtensorflow_ENABLE_GPU=OFF
But, with Dtensorflow_ENABLE_GPU=ON, I am getting the following error

"D:\temp\tensorflow\tensorflow\contrib\cmake\build\tf_python_build_pip_package.vcxproj" (default target) (1) ->
"D:\temp\tensorflow\tensorflow\contrib\cmake\build\pywrap_tensorflow.vcxproj" (default target) (3) ->
"D:\temp\tensorflow\tensorflow\contrib\cmake\build\tf_core_gpu_kernels.vcxproj" (default target) (33) ->
(CustomBuild target) ->
  C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V140\Microsoft.CppCommon.targets(171,5): error MSB6006: "cmd.exe" e
xited with code 1. [D:\temp\tensorflow\tensorflow\contrib\cmake\build\tf_core_gpu_kernels.vcxproj]

    362 Warning(s)
    1 Error(s)

Any idea how to fix this

Thanks

Carmezim commented:
@karthiek 
This issue seems to be related to [this](https://github.com/tensorflow/tensorflow/pull/5071#issuecomment-255236576).

guschmue commented:
Looks it fails compiling the cuda kernels. The real error messages is earlier in the output. If you can get that for me ...

Carmezim commented:
Had similar (kernels) problems:
`"c:\...\tensorflow\tensorflow\contrib\cmake\build\tf_tutorials_example_trainer.vcxproj" (default ta
rget) (1) ->
"C:\...\tensorflow\tensorflow\contrib\cmake\build\tf_core_gpu_kernels.vcxproj" (default target) (81
) ->
(CustomBuild target) ->
  C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V140\Microsoft.CppCommon.targets(171,5): error MSB6006: "cmd.exe" e
xited with code 1. [C:\...\tensorflow\tensorflow\contrib\cmake\build\tf_core_gpu_kernels.vcxproj]`

Do you need the warnings?
```
c:\...\tensorflow\tensorflow\core\kernels\matrix_band_part_op.cc(175): warning C4506: no definition
 for inline function 'void tensorflow::functor::MatrixBandPart<tensorflow::GPUDevice,float>::Compute(const Device &,Eig
en::DenseIndex,Eigen::DenseIndex,Eigen::TensorMap<Eigen::Tensor<const T,3,1,IndexType>,16>,Eigen::TensorMap<Eigen::Tens
or<T,3,1,IndexType>,16>)' [C:\Users\Adriano\Documents\tensorflow\tensorflow\contrib\cmake\build\tf_core_kernels.vcxproj
]
          with
          [
              Device=tensorflow::GPUDevice,
              T=float,
              IndexType=Eigen::DenseIndex
          ]
```
This is one of the many similar warnings related to the eigen patch you mentioned.

Had plenty of this
```
 c:\...\tensorflow\tensorflow\core\kernels\sparse_tensor_dense_matmul_op.cc(283): warning C4506: n
o definition for inline function 'void tensorflow::functor::SparseTensorDenseMatMulFunctor<tensorflow::GPUDevice,float,
false,true>::Compute(const Device &,Eigen::TensorMap<Eigen::Tensor<T,2,1,IndexType>,16>,Eigen::TensorMap<Eigen::Tensor<
const __int64,2,1,IndexType>,16>,Eigen::TensorMap<Eigen::Tensor<const T,1,1,IndexType>,16>,Eigen::TensorMap<Eigen::Tens
or<const T,2,1,IndexType>,16>,Eigen::TensorMap<Eigen::Tensor<T,1,1,IndexType>,16>)' [C:\...\tensorf
low\tensorflow\contrib\cmake\build\tf_core_kernels.vcxproj]
```
And at last before the error message:
```
(ClCompile target) ->
  c:\users\adriano\documents\tensorflow\tensorflow\stream_executor\lib\statusor.h(190): warning C4180: qualifier applie
d to function type has no meaning; ignored (compiling source file C:\Users\Adriano\Documents\tensorflow\tensorflow\stre
am_executor\plugin_registry.cc) [C:\Users\Adriano\Documents\tensorflow\tensorflow\contrib\cmake\build\tf_stream_executo
r.vcxproj]
  c:\users\adriano\documents\tensorflow\tensorflow\stream_executor\cuda\cuda_driver.cc(233): warning C4068: unknown pra
gma [C:\Users\Adriano\Documents\tensorflow\tensorflow\contrib\cmake\build\tf_stream_executor.vcxproj]
  c:\users\adriano\documents\tensorflow\tensorflow\stream_executor\cuda\cuda_driver.cc(234): warning C4068: unknown pra
gma [C:\Users\Adriano\Documents\tensorflow\tensorflow\contrib\cmake\build\tf_stream_executor.vcxproj]
  c:\users\adriano\documents\tensorflow\tensorflow\stream_executor\cuda\cuda_driver.cc(320): warning C4068: unknown pra
gma [C:\Users\Adriano\Documents\tensorflow\tensorflow\contrib\cmake\build\tf_stream_executor.vcxproj]
```:

guschmue commented:
the warnings will not prevent the build from completing. Somewhere between the warnings should be a error message in the format 'somesource.cc(line#): error ....'. 
Maybe search from the top of the output for ': error'

guschmue commented:
line 888 in your output: nvcc fatal   : Stray '"' character
Might be that one of the -D options that cmake is called with has some wrong quote.
Best way to debug would be vi tf_core_gpu_kernels_generated_adjust_contrast_op_gpu.cu.cc.obj.RelWithDebInfo.cmake

CamJohnson26 commented:
Hi I believe I had the same error, cmake bat file looked like so:
cd "C:\Users\Cameron\Desktop\tensorflow\tensorflow\tensorflow\contrib\cmake\build"
cmake .. -A x64 -DCMAKE_BUILD_TYPE=Release ^
-DSWIG_EXECUTABLE=C:/swigwin-3.0.10/swig.exe ^
-DPYTHON_EXECUTABLE=C:/Python35/python.exe ^
-DPYTHON_LIBRARIES=C:/Python35/libs/python35.lib ^
-Dtensorflow_ENABLE_GPU=ON ^
-DCUDNN_HOME=C:/cuda

karthiek commented:
@guschmue 
I guess there is a problem with the line 81 in  tf_core_gpu_kernels_generated_adjust_contrast_op_gpu.cu.cc.obj.RelWithDebInfo.cmake
the line is 
set(CUDA_NVCC_COMPILE_DEFINITIONS "EIGEN_AVOID_STL_ARRAY;NOMINMAX;_WIN32_WINNT=0x0A00;LANG_CXX11;COMPILER_MSVC;__VERSION__="MSVC";WIN32;OS_WIN;_MBCS;WIN64;WIN32_LEAN_AND_MEAN;NOGDI;PLATFORM_WINDOWS;TENSORFLOW_USE_EIGEN_THREADPOOL;EIGEN_HAS_C99_MATH;_ITERATOR_DEBUG_LEVEL=0;GOOGLE_CUDA=1;TF_EXTRA_CUDA_CAPABILITIES=3.5,5.2") # list (needs to be in quotes to handle spaces properly).

when I print the value of CUDA_NVCC_COMPILE_DEFINITIONS , I get the following:
EIGEN_AVOID_STL_ARRAY;NOMINMAX;_WIN32_WINNT=0x0A00;LANG_CXX11;**COMPILER_MSVC;__VERSION__=;MSVC";WIN32**;OS_WIN;_MBCS;WIN64;WIN32_LEAN_AND_MEAN;NOGDI;PLATFORM_WINDOWS;TENSORFLOW_USE_EIGEN_THREADPOOL;EIGEN_HAS_C99_MATH;_ITERATOR_DEBUG_LEVEL=0;GOOGLE_CUDA=1;TF_EXTRA_CUDA_CAPABILITIES=3.5,5.2"

Is the part highlighted expected to be that way. I guess the stray character referred belongs to this line.

CamJohnson26 commented:
Is it a problem with spaces in the CMakeCache.txt file? I have my CUDA files installed in Program FIles and the path isn't quote wrapped in CMakeCache.txt

guschmue commented:
going to look at it shortly. My cuda is installed under program files as well and its happy. Definitely the "MSVC" looks suspicious, just need to understand why some people have the problem and others don't.
What version of cmake do you use ?

CamJohnson26 commented:
I'm on 3.7.0-rc2, downloaded yesterday

karthiek commented:
I'm on 3.7.0-rc2 too.

guschmue commented:
ok, I'm on 3.5.2 and 3.6 - let me update and see my build breaks too.

karthiek commented:
I tried with cmake 3.6 version and was able to build successfully.

guschmue commented:
thank you, knowing that its something in cmake 3.7 helps a ton. Going see if we can do something on it.

guschmue commented:
I had internally setup a jenkins build for windows gpu from tf master which is unhappy with https://github.com/tensorflow/tensorflow/pull/5127. Going to look at it today.

Carmezim commented:
I had now these kind of errors building with cmake 3.6. Not having much luck.
```
"c:\tensorflow\tensorflow\contrib\cmake\build\tf_tutorials_example_trainer.vcxproj" (default ta
rget) (1) ->
"C:\tensorflow\tensorflow\contrib\cmake\build\tf_core_gpu_kernels.vcxproj" (default target) (81
) ->
(CustomBuild target) ->
  C:/tensorflow/tensorflow/core/kernels/cwise_op_gpu_select.cu.cc(42): error : namespace "Eigen
" has no member "IndexList" [C:\tensorflow\tensorflow\contrib\cmake\build\tf_core_gpu_kernels.v
cxproj]
  C:/tensorflow/tensorflow/core/kernels/cwise_op_gpu_select.cu.cc(42): error : namespace "Eigen
" has no member "type2index" [C:\tensorflow\tensorflow\contrib\cmake\build\tf_core_gpu_kernels.
vcxproj]
```
```

 1844 Warning(s)
    21 Error(s)
```

CamJohnson26 commented:
Same exact problem with 3.6. Looks like a problem with Eigen, not sure if it matters that I'm using vs 2015 community, not enterprise? Don't have the exact error but can post at home

guschmue commented:
this is something that came in with #5127. You can do a 
git reset --hard 41ba1e0e6fbf443f1a972ff0130ba6741b9b7a50
to go before that change.

Carmezim commented:
@guschmue will do. Thank you.

CamJohnson26 commented:
Still getting build errors even after resetting the branch, this time after 4 hours 30 minutes. Error is 

> "C:\Users\Cameron\Desktop\tensorflow\tensorflow\tensorflow\contrib\cmake\build\tf_python_build_pip_package.vcxproj" (default
> target) (1) ->
> (PostBuildEvent target) ->
>   EXEC : error : invalid command 'bdist_wheel' [C:\Users\Cameron\Desktop\tensorflow\tensorflow\tensorflow\contrib\cmake\build
> \tf_python_build_pip_package.vcxproj]
>   C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V140\Microsoft.CppCommon.targets(133,5): error MSB3073: The command "setl
> ocal\r [C:\Users\Cameron\Desktop\tensorflow\tensorflow\tensorflow\contrib\cmake\build\tf_python_build_pip_package.vcxproj]
> C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V140\Microsoft.CppCommon.targets(133,5): error MSB3073: "C:\Program Files\C
> Make\bin\cmake.exe" -E copy C:/Users/Cameron/Desktop/tensorflow/tensorflow/tensorflow/contrib/cmake/setup.py C:/Users/Cameron
> /Desktop/tensorflow/tensorflow/tensorflow/contrib/cmake/build/tf_python/\r [C:\Users\Cameron\Desktop\tensorflow\tensorflow\te
> nsorflow\contrib\cmake\build\tf_python_build_pip_package.vcxproj]
> C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V140\Microsoft.CppCommon.targets(133,5): error MSB3073: if %errorlevel% neq

...

>     2948 Warning(s)
>     2 Error(s)

CamJohnson26 commented:
"pip install wheel" fixed this, finally have a successful build.

guschmue commented:
Cool. I have https://github.com/tensorflow/tensorflow/pull/5157 out to fix the master for windows gpu as well.

CamJohnson26 commented:
Another note, after installing with pip I also had to set the LD_LIBRARY_PATH environment variable to C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v8.0\lib\Win32. May want to add these steps to the readme. Thanks btw for this feature, been excited about it for a while

guschmue commented:
Hm, that is odd: the nvidia dlls need to be in PATH but the nvidia installer adds them to PATH already (I forgot if this is maybe optional). The dll that is not in the PATH is the cudnn dll because that comes from a zip file. So you are right, we should add something to the README.md. On windows LD_LIBRARY_PATH is odd, maybe using PATH is more the windows way.
I polish the readme some tomorrow.
Thank you for trying this out - greatly appreciated and invaluable to have more eyeballs!!!
Btw, there are a few smaller issues we know about - we'll address them in a PR sometime this week.

guschmue commented:
Sure, please fix the readme.
Good to know that it works for you – thank you for verifying! There are a few minor issues we know about but we have another pr coming up that will address those.

From: Adriano Carmezim [mailto:notifications@github.com]
Sent: Thursday, October 20, 2016 5:48 PM
To: tensorflow/tensorflow <tensorflow@noreply.github.com>
Cc: Guenther Schmuelling <guschmue@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [tensorflow/tensorflow] support for gpu on windows (#5071)


@guschmue<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fguschmue&data=02%7C01%7Cguschmue%40microsoft.com%7C48e38779b02c466c1f0a08d3f94bef29%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636126076941600834&sdata=pi6j18KwNKXJmziDmYUOvuVBym8DozZ%2BYyBKivHi1pE%3D&reserved=0> All working great so far, thank you for this contribution.
If you want I can make those edits on "current limitations" and "current status" regarding the availability for GPU support and submit a PR, just ping me if you will.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Ftensorflow%2Ftensorflow%2Fpull%2F5071%23issuecomment-255265994&data=02%7C01%7Cguschmue%40microsoft.com%7C48e38779b02c466c1f0a08d3f94bef29%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636126076941600834&sdata=jecyqg1wOurTyR7O6s4odF8oQwQm5B7TTJxLAGa3ux8%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAV4NiA_xiknMX7FJ2Q54DDKxRtoI73kaks5q2AvKgaJpZM4KbRfF&data=02%7C01%7Cguschmue%40microsoft.com%7C48e38779b02c466c1f0a08d3f94bef29%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636126076941600834&sdata=otJznv2FyAE5saFEDRR6Ak8JSzIX92us%2FuhfOFDKAKg%3D&reserved=0>.

Carmezim commented:
@guschmue what?

Carmezim commented:
@guschmue was this all supposed to show here?

> From: Adriano Carmezim [mailto:notifications@github.com]
> Sent: Thursday, October 20, 2016 5:48 PM
> To: tensorflow/tensorflow <tensorflow@noreply.github.com>
> Cc: Guenther Schmuelling <guschmue@microsoft.com>; Mention <mention@noreply.github.com>
> Subject: Re: [tensorflow/tensorflow] support for gpu on windows (#5071)
> 
> 
> @guschmue<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fguschmue&data=02%7C01%7Cguschmue%40microsoft.com%7C48e38779b02c466c1f0a08d3f94bef29%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636126076941600834&sdata=pi6j18KwNKXJmziDmYUOvuVBym8DozZ%2BYyBKivHi1pE%3D&reserved=0> All working great so far, thank you for this contribution.
> If you want I can make those edits on "current limitations" and "current status" regarding the availability for GPU support and submit a PR, just ping me if you will.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Ftensorflow%2Ftensorflow%2Fpull%2F5071%23issuecomment-255265994&data=02%7C01%7Cguschmue%40microsoft.com%7C48e38779b02c466c1f0a08d3f94bef29%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636126076941600834&sdata=jecyqg1wOurTyR7O6s4odF8oQwQm5B7TTJxLAGa3ux8%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAV4NiA_xiknMX7FJ2Q54DDKxRtoI73kaks5q2AvKgaJpZM4KbRfF&data=02%7C01%7Cguschmue%40microsoft.com%7C48e38779b02c466c1f0a08d3f94bef29%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636126076941600834&sdata=otJznv2FyAE5saFEDRR6Ak8JSzIX92us%2FuhfOFDKAKg%3D&reserved=0>.

knownsoldier commented:
@guschmue 
I'm getting an NVCC error while trying to compile with GPU support. My setup is:
* Windows 10
* Microsoft Visual Studio Community 2015 with Visual C++ 2015
* CUDNN 5.1
* CUDA 8.0
* Anaconda3 4.2.0
* swigwin-3.0.10

Relevant environment variables:
```
CUDA_PATH=C:\NVIDIAGPUComputingToolkit\CUDA\v8.0
CUDA_PATH_V8_0=C:\NVIDIAGPUComputingToolkit\CUDA\v8.0
Path=C:\NVIDIAGPUComputingToolkit\CUDA\v8.0\bin;C:\NVIDIAGPUComputingToolkit\CUDA\v8.0\libnvvp;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Program Files (x86)\Skype\Phone\;C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;C:\Program Files\Git\cmd;C:\Users\<user>\.dnx\bin;C:\Program Files\Microsoft DNX\Dnvm\;C:\Program Files\Microsoft SQL Server\130\Tools\Binn\;C:\Program Files (x86)\Windows Kits\8.1\Windows Performance Toolkit\;C:\Users\<user>\cudnn\bin\;C:\Users\<user>\Anaconda3;C:\Users\<user>\Anaconda3\Scripts;C:\Users\<user>\Anaconda3\Library\bin;C:\Users\<user>\AppData\Local\Microsoft\WindowsApps;C:\Python;C:\Program Files\CMake\bin;C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\
```
I'm running the following CMake command standing in `tensorflow\contrib\cmake\build\` without problems:
```
cmake .. -A x64 -DCMAKE_BUILD_TYPE=Release -DSWIG_EXECUTABLE=C:\Users\<user>\swigwin-3.0.10\swig.exe -DPYTHON_EXECUTABLE=C:\Users\<user>\Anaconda3\python.exe -DPYTHON_LIBRARIES=C:\Users\<user>\Anaconda3\libs\python35.lib -DCUDA_TOOLKIT_ROOT_DIR=C:\NVIDIAGPUComputingToolkit\CUDA\v8.0  -Dtensorflow_ENABLE_GPU=ON -DCUDNN_HOME=C:\Users\<user>\cudnn\
```
I run the following MSBuild command:
```
C:\Windows\Microsoft.NET\Framework64\v4.0.30319\MSBuild /p:Configuration=Release tf_tutorials_example_trainer.vcxproj
```
It runs for a while and terminates with the following error:
```
Project "C:\Users\<user>\tensorflow\tensorflow\contrib\cmake\build\tf_tutorials_example_trainer.vcxproj" (1) is building "C:\Users\<user>\tensorflow
\tensorflow\contrib\cmake\build\tf_core_gpu_kernels.vcxproj" (81) on node 1 (default targets).
InitializeBuildStatus:
  Touching "tf_core_gpu_kernels.dir\Release\tf_core_.99484E5D.tlog\unsuccessfulbuild".
CustomBuild:
  Building NVCC (Device) object CMakeFiles/tf_core_gpu_kernels.dir/__/__/core/kernels/Release/tf_core_gpu_kernels_generated_adjust_contrast_op_gpu.cu.cc.obj
  nvcc fatal   : A single input file is required for a non-link phase when an outputfile is specified

  CMake Error at tf_core_gpu_kernels_generated_adjust_contrast_op_gpu.cu.cc.obj.Release.cmake:207 (message):
    Error generating
    C:/Users/<user>/tensorflow/tensorflow/contrib/cmake/build/CMakeFiles/tf_core_gpu_kernels.dir/__/__/core/kernels/Release/tf_core_gpu_kernels_generated_adjust_contrast_op_gpu.cu.cc.obj


C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V140\Microsoft.CppCommon.targets(171,5): error MSB6006: "cmd.exe" exited with code 1. [C:\Users\<user>\tensorflow\tensorflow\contrib\cmake\build\tf_core_gpu_kernels.vcxproj]
Done Building Project "C:\Users\<user>\tensorflow\tensorflow\contrib\cmake\build\tf_core_gpu_kernels.vcxproj" (default targets) -- FAILED.
.
.
.
"C:\Users\<user>\tensorflow\tensorflow\contrib\cmake\build\tf_tutorials_example_trainer.vcxproj" (default target) (1) ->
"C:\Users\<user>\tensorflow\tensorflow\contrib\cmake\build\tf_core_gpu_kernels.vcxproj" (default target) (81) ->
(CustomBuild target) ->
  C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V140\Microsoft.CppCommon.targets(171,5): error MSB6006: "cmd.exe" exited with code 1. [C:\Users\<user>\tensorflow\tensorflow\contrib\cmake\build\tf_core_gpu_kernels.vcxproj]

    5 Warning(s)
    1 Error(s)
```

From what I can tell I've followed all instructions from `tensorflow\tensorflow\contrib\cmake\README.md`. What am I missing?

Carmezim commented:
Isn't it the Cmake version?
